File: router/internal/providers/adapters/agentrouter.go
package adapters

import "net/http"

// agentRouterSurfaces is the single surface the AgentRouter adapter serves.
var agentRouterSurfaces = []Surface{SurfaceAnthropicMessages}

// agentRouterModel is the lone model the AgentRouter adapter publishes.
func agentRouterModel() ProviderModel {
	caps := ProviderCaps{
		Surfaces:  append([]Surface(nil), agentRouterSurfaces...),
		Streaming: true,
		Reasoning: true,
		Images:    true,
	}
	return Model("claude-opus-4-8", "Claude Opus 4.8", &caps)
}

// AgentRouterAdapter speaks native Anthropic Messages behind a client-
// identity gate. Its provider-specific endpoint, identity headers, and
// classification remain in the adapter; account selection, retries, and
// credential lifecycle stay in their respective runtime owners.
type AgentRouterAdapter struct {
	meta    ProviderMeta
	caps    ProviderCaps
	catalog *staticCatalog
}

// AgentRouterAdapterConfig carries provider-owned identity and catalog data.
// It deliberately contains no account-selection or retry settings.
type AgentRouterAdapterConfig struct {
	ID             string
	DisplayName    string
	BaseURL        string
	CredentialRef  string
	CredentialKind CredentialKind
	Models         []ProviderModel
}

// NewAgentRouterAdapter returns the AgentRouter provider.
func NewAgentRouterAdapter(configs ...AgentRouterAdapterConfig) *AgentRouterAdapter {
	var cfg AgentRouterAdapterConfig
	if len(configs) > 0 {
		cfg = configs[0]
	}
	if cfg.ID == "" {
		cfg.ID = "agentrouter"
	}
	if cfg.DisplayName == "" {
		cfg.DisplayName = "AgentRouter"
	}
	if cfg.BaseURL == "" {
		cfg.BaseURL = "https://agentrouter.org"
	}
	if cfg.CredentialKind == "" {
		cfg.CredentialKind = CredentialAPIKey
	}
	if cfg.CredentialRef == "" {
		cfg.CredentialRef = providerCredentialRef(cfg.ID)
	}
	models := append([]ProviderModel(nil), cfg.Models...)
	if len(models) == 0 {
		models = []ProviderModel{agentRouterModel()}
	}
	caps := aggregateCapabilities(models, ProviderCaps{
		Surfaces:  append([]Surface(nil), agentRouterSurfaces...),
		Streaming: true,
		Reasoning: true,
		Images:    true,
	})
	return &AgentRouterAdapter{
		meta: ProviderMeta{
			ID:              cfg.ID,
			DisplayName:     cfg.DisplayName,
			Protocol:        ProtocolAnthropic,
			CredentialKind:  cfg.CredentialKind,
			CredentialKinds: []CredentialKind{cfg.CredentialKind},
			CredentialRef:   cfg.CredentialRef,
			BaseURL:         cfg.BaseURL,
		},
		caps:    caps,
		catalog: newStaticCatalog(models),
	}
}

func (p *AgentRouterAdapter) Metadata() ProviderMeta       { return p.meta }
func (p *AgentRouterAdapter) Capabilities() ProviderCaps   { return p.caps }
func (p *AgentRouterAdapter) Models() ProviderModelCatalog { return p.catalog }

func (p *AgentRouterAdapter) ResolveTarget(modelID string, surface Surface) (RouteTarget, error) {
	if surface != SurfaceAnthropicMessages {
		return RouteTarget{}, &UnknownSurfaceError{ProviderID: p.meta.ID, Surface: surface}
	}
	entry := p.catalog.Get(modelID)
	if entry == nil {
		return RouteTarget{}, &UnknownModelError{ProviderID: p.meta.ID, ModelID: modelID}
	}
	return RouteTarget{
		ProviderID:      p.meta.ID,
		ModelID:         entry.ID,
		UpstreamModelID: entry.UpstreamID,
		Surface:         surface,
	}, nil
}

func (p *AgentRouterAdapter) Endpoint(target RouteTarget) Endpoint {
	return Endpoint{
		Method: http.MethodPost,
		Path:   "v1/messages",
		Query:  map[string]string{"beta": "true"},
	}
}

func (p *AgentRouterAdapter) AuthMaterial(credential string, target RouteTarget) (AuthMaterial, error) {
	if target.ProviderID != "" && target.ProviderID != p.meta.ID {
		return AuthMaterial{}, &IDMismatchError{AdapterID: p.meta.ID, TargetID: target.ProviderID}
	}
	if target.Surface != SurfaceAnthropicMessages {
		return AuthMaterial{}, &UnknownSurfaceError{ProviderID: p.meta.ID, Surface: target.Surface}
	}
	if credential == "" {
		return AuthMaterial{}, &AuthError{ProviderID: p.meta.ID, Reason: "AgentRouter requires an API key"}
	}
	headers := http.Header{}
	headers.Set("content-type", "application/json")
	headers.Set("anthropic-version", "2023-06-01")
	headers.Set("anthropic-beta", "claude-code-20250219,interleaved-thinking-2025-05-14,effort-2025-11-24")
	headers.Set("anthropic-dangerous-direct-browser-access", "true")
	headers.Set("x-app", "cli")
	headers.Set("user-agent", "claude-cli/2.1.195 (external, sdk-cli)")
	headers.Set("x-claude-code-session-id", randomUUID())
	headers.Set("x-stainless-retry-count", "0")
	headers.Set("x-stainless-timeout", "600")
	headers.Set("x-stainless-lang", "js")
	headers.Set("x-stainless-package-version", "0.94.0")
	headers.Set("x-stainless-os", "MacOS")
	headers.Set("x-stainless-arch", "arm64")
	headers.Set("x-stainless-runtime", "node")
	headers.Set("x-stainless-runtime-version", "v24.3.0")
	headers.Set("accept-encoding", "gzip, deflate, br, zstd")
	headers.Set("x-api-key", credential)
	return AuthMaterial{Headers: headers}, nil
}

func (p *AgentRouterAdapter) BuildRequest(envelope RequestEnvelope, credential string) (BuiltRequest, error) {
	auth, err := p.AuthMaterial(credential, envelope.Target)
	if err != nil {
		return BuiltRequest{}, err
	}
	return BuiltRequest{
		Endpoint: p.Endpoint(envelope.Target),
		Body:     envelope.Body,
		Auth:     auth,
		Stream:   envelope.Stream,
	}, nil
}

func (p *AgentRouterAdapter) ClassifyResponse(evidence ResponseEvidence) ClassifiedResponse {
	return classifyByStatus(evidence)
}

File: src/providers/integrations/agentrouter.ts
// Bespoke by wire protocol (Phase C4): the AgentRouter Claude CLI
// fingerprint — the `anthropic-beta` list, the Stainless identity headers,
// `x-app: cli`, and a per-request `x-claude-code-session-id` — is layered on
// top of the shared Claude Messages request. Projection, URL resolution, SSE
// decoding, and stop-reason mapping all come from the canonical pieces
// (`buildClaudeMessagesRequest` + `sendClaudeMessagesRequest`), matching the
// kimi and claude-code Messages adapters.
//
// The fingerprint cannot ride on the builder's `customHeaders`:
// `CLAUDE_EXTRA_PROTECTED_HEADERS` rejects `anthropic-beta`, `x-app`, the
// `x-stainless-*` family, `x-claude-code-session-id`, and `anthropic-version`
// for operator-supplied headers, so they are layered onto the built headers
// at this call site instead.
import { GatewayError } from "../../transport/gateway-error";
import type { CanonicalEvent, CanonicalRequest } from "../../transport/canonical-model";
import { readCredentialSecret, type ProviderDispatchTarget, type ProviderAdapter, type ProviderDispatchContext } from "../provider-registry";
import { providerBaseUrl } from "../provider-metadata";
import { buildClaudeMessagesRequest } from "./claude-messages";
import { sendClaudeMessagesRequest } from "../../protocol/transport/messages";

export const AGENTROUTER_PROVIDER_ID = "agentrouter" as const;
export const AGENTROUTER_BASE_URL = `${providerBaseUrl("agentrouter")}/v1/messages?beta=true`;

/**
 * The AgentRouter CLI fingerprint headers. Every value is part of the
 * provider's wire identity and must stay byte-identical. `content-type`,
 * `anthropic-version`, and `x-api-key` are already produced by
 * `buildClaudeMessagesRequest` with these same values and are therefore not
 * repeated here.
 */
function agentRouterFingerprintHeaders(stream: boolean): Record<string, string> {
  return {
    "anthropic-beta": "claude-code-20250219,interleaved-thinking-2025-05-14,effort-2025-11-24",
    "anthropic-dangerous-direct-browser-access": "true",
    "x-app": "cli",
    "user-agent": "claude-cli/2.1.195 (external, sdk-cli)",
    "x-claude-code-session-id": crypto.randomUUID(),
    "x-stainless-retry-count": "0",
    "x-stainless-timeout": "600",
    "x-stainless-lang": "js",
    "x-stainless-package-version": "0.94.0",
    "x-stainless-os": "MacOS",
    "x-stainless-arch": "arm64",
    "x-stainless-runtime": "node",
    "x-stainless-runtime-version": "v24.3.0",
    accept: stream ? "text/event-stream" : "application/json",
    "accept-encoding": "gzip, deflate, br, zstd",
  };
}

// AgentRouter model catalog.

import { defineModel } from "../model-definition";
import type { ModelDefinition } from "../provider-registry";

export const AGENTROUTER_MODELS: readonly ModelDefinition[] = [
  defineModel({
    id: "claude-sonnet-4-5",
    wireFamily: "messages",
    endpoint: "/v1/messages?beta=true",
    ctx: 200_000,
    out: 64_000,
    vision: true,
    reasoning: true,
    toolCall: true,
    webSearch: false,
  }),
];

interface AgentRouterAdapterOptions {
  readonly fetch?: typeof fetch;
  readonly baseUrl?: string;
}

class AgentRouterAdapter implements ProviderAdapter {
  readonly provider_id = AGENTROUTER_PROVIDER_ID;
  private readonly fetchFn: typeof fetch;
  private readonly baseUrl: string;

  constructor(options: AgentRouterAdapterOptions = {}) {
    this.fetchFn = options.fetch ?? globalThis.fetch;
    this.baseUrl = options.baseUrl ?? AGENTROUTER_BASE_URL;
  }

  async *dispatch(
    request: CanonicalRequest,
    candidate: ProviderDispatchTarget,
    context: ProviderDispatchContext,
  ): AsyncIterable<CanonicalEvent> {
    if (candidate.wire_family !== "messages") {
      throw new GatewayError("capability_unsupported", 400, `agentrouter supports messages only, got ${candidate.wire_family}`);
    }
    const apiKey = readCredentialSecret(context.credential, "AgentRouter API key is required");

    // Endpoint resolution stays provider-specific: an absolute candidate path
    // wins, a non-default path is appended to the origin with the Messages
    // beta, and the default keeps the `?beta=true` base URL. The resolved
    // absolute URL is handed to the shared builder, whose `endpointUrl`
    // returns absolute paths verbatim.
    const url =
      candidate.endpoint_path && candidate.endpoint_path.startsWith("http")
        ? candidate.endpoint_path
        : candidate.endpoint_path && candidate.endpoint_path.length > 0 && candidate.endpoint_path !== "/v1/messages" && candidate.endpoint_path !== "/v1/messages?beta=true"
          ? `${this.baseUrl.replace(/\?beta=true$/, "")}${candidate.endpoint_path.startsWith("/") ? candidate.endpoint_path : `/${candidate.endpoint_path}`}?beta=true`
          : this.baseUrl;

    const outbound = buildClaudeMessagesRequest({
      request,
      authHeader: "x-api-key",
      credential: { secret: apiKey },
      endpointPath: url,
      mutatePayload: (payload) => {
        payload["model"] = candidate.model_id || request.model;
        return payload;
      },
    });
    Object.assign(outbound.headers, agentRouterFingerprintHeaders(request.stream));

    yield* sendClaudeMessagesRequest(
      outbound.url,
      outbound.headers,
      outbound.body,
      context,
      request,
      this.fetchFn,
    );
  }
}

export function createAgentRouterAdapter(options: AgentRouterAdapterOptions = {}): ProviderAdapter {
  return new AgentRouterAdapter(options);
}
